### DISCUSSION TITLE
Question on controller frequency

### DISCUSSION THREAD

**Discussion by StefanoS90** [eleurent/highway-env#170] [General]
Hello! and thanks for the great work.

One curiosity about the frequency of the vehicle control:
From what i see in the _simulate() function, an high level action (e.g. LANE_LEFT/LANE_RIGHT) is performed at a policy frequency and kept for a certain number of simulation steps. 
If i understood correctly, the low level steering and speed controller are also updated at a policy frequency. Wouldn't be better to make them faster compared to the high level decision?

**Comment by mhtb32** [comment]
Maybe #63 helps you.

**Comment by eleurent** [comment]
>From what i see in the _simulate() function, an high level action (e.g. LANE_LEFT/LANE_RIGHT) is performed at a policy frequency and kept for a certain number of simulation steps.

That's right.

> If i understood correctly, the low level steering and speed controller are also updated at a policy frequency. Wouldn't be better to make them faster compared to the high level decision?

No, low-level steering and speed control is performed at simulation frequency (higher than policy/decision). I can imagine that it's not really obvious from reading the code, but here's how it works:

1. In MDPVehicle.act(action) when the action is None, the current speed/lane targets are unchanged, and the action is forwarded to the ControlledVehicle.act() method, which computes the steering and acceleration controls.
2. ego_vehicle.act(action) can thus be called at policy frequency for high-level decisions, but ego.vehicle.act() also needs to be called at simulation frequency for control.
3. this is achieved in the _simulate() function by calling road.act(), which itself broadcasts vehicle.act() to every vehicle in the scene.

**Comment by StefanoS90** [comment]
Thanks for the answer. I did not see that ego.vehicle.act() was called also in road.act().
